tests/test_esp_generator.py

```diff
@@ -0,0 +1,414 @@
+
+    return _mock_system_data
+
+
+@contextmanager
+def loader_device_part_uuid(value, root):
+    """LoaderDevicePartUUID EFI variable context manager"""
+    varpath = root / 'sys/firmware/efi/efivars' / espgen.LOADER_DEVICE_PART_UUID_EFIVAR
+    varpath.parent.mkdir(parents=True, exist_ok=True)
+
+    # Using the utf-16 codec for writing will add the byte order mark
+    # (BOM). Select the native endian version.
+    utf_16 = 'utf-16-le' if sys.byteorder == 'little' else 'utf-16-be'
+
+    with open(varpath, 'wb') as f:
+        f.write(b'\0\0\0\0')
+        f.write(value.encode(utf_16))
+
+        # Add 3 nul byte terminators like systemd-boot does.
+        f.write(b'\0\0\0')
+
+    try:
+        yield
+    finally:
+        varpath.unlink()
+
+
+# Dictionary of mocked data to use with the test_get_esp_mount() test. The keys
+# are names and the values are expected EspMount instances.
+ESP_MOUNT_TEST_DATA = {
+    # Standard EOS using grub with GPT partitioning. /boot gets mounted in the
+    # initramfs by ostree-prepare-root.
+    'grub-gpt': espgen.EspMount(
+        source='/dev/vda1',
+        target='/efi',
+        type='vfat',
+        umask='0077',
+    ),
+
+    # Standard EOS using grub with MBR partitioning. /boot gets mounted in the
+    # initramfs by ostree-prepare-root.
+    'grub-mbr': espgen.EspMount(
+        source='/dev/vda1',
+        target='/efi',
+        type='vfat',
+        umask='0077',
+    ),
+
+    # Standard EOS using grub with GPT partitioning and /efi in fstab. /boot
+    # gets mounted in the initramfs by ostree-prepare-root. It's not typical
+    # that there would be a /efi entry in fstab, but this test ensures no /efi
+    # mount units are created.
+    'grub-gpt-fstab-efi': None,
+
+    # EOS using systemd-boot with GPT partitioning. This represents PAYG
+    # systemd. There is no /boot mounted by ostree-prepare-root since the root
+    # partition's /boot directory is empty. The ESP should be mounted at /boot
+    # world readable so that ostree deployments can be resolved unprivileged.
+    'sdboot': espgen.EspMount(
+        source='/dev/vda1',
+        target='/boot',
+        type='vfat',
+        umask='0022',
+    ),
+
+    # EOS using systemd-boot with GPT partitioning and an XBOOTLDR partition.
+    # This is not something we currently do or try to support since it would
+    # provide little benefit. Mostly this is here to ensure we don't add
+    # incorrect mounts if the situation starts happening.
+    'sdboot-xbootldr': None,
+
+    # Windows dual boot. The Endless disk (including an ESP partition) is a
+    # file on the NTFS partition that gets loop mounted. The real disk's ESP
+    # should be mounted instead of the loop disk's ESP partition. The real disk
+    # is found from the endless.image.device kernel command line parameter.
+    'windows': espgen.EspMount(
+        source='/dev/sda1',
+        target='/efi',
+        type='vfat',
+        umask='0077',
+    ),
```

Comment:
There are 2 Windows dual-boot cases: efi and bios. In the bios case we don't expect to find an mbr. Perhaps we should rename the test case here as `windows-efi` so that in future some brave explorer can find a BIOS windows system in a cave, install EOS on it and check that in that case the generator doesn't break.

Reply:
Good point, I agree with the rename. I think I was trying to support the idea that the ESP should be mounted whether you're booted with UEFI or not. However, thinking about it again, I don't think that's a great idea. Especially since part of the method depends on reading EFI variables, which obviously won't be there. I think I'll add a short circuit if you're not in UEFI mode.

Reply:
On dual-boot systems (unlike standalone ones) we only ever install the corresponding bootloader for the firmware type, so if a dual-boot system is booted in BIOS mode then it is certain that there is no EFI bootloader.

In theory I guess it's possible that a standalone install may be booted in CSM/legacy BIOS mode one day and real EFI mode another day if someone is doing weird stuff with the firmware settings. In that case I think it's still fine to short-circuit if booted in BIOS mode and not mount the ESP (which ultimately we're doing in order to update the bootloader) even if there is one. If the system is later booted in EFI mode we'll have a chance to update it then; and if not then it's a moot point whether the EFI bootloaders are out of date.

Reply:
Did the rename and added a `/sys/firmware/efi` check (which is also how `systemd-gpt-auto-generator` decides if you're EFI booted or not).